[docs] trustless swap updates (#19409)
MystenLabs/sui: docs/content/guides/developer/app-examples/trustless-swap.mdx

After a trade is initiated, you don't want the trading party to modify the object they agreed to trade. Imagine you're trading in-game items and you agree to trade a weapon with all its attachments, and its owner strips all its attachments just before the trade.

In a traditional trade, a third party typically holds the items in escrow to make sure they are not tampered with before the trade completes. This requires either trust that the third party won't tamper with it themselves, paying the third party to ensure that doesn't happen, or both.

In a trustless swap, however, you can use the safety properties of Move to force an item's owner to prove that they have not tampered with the version of the object that you agreed to trade, without involving anyone else.

This is done by requiring that an object that is available for trading is **locked** with a **single-use key**, and asking the owner to supply the key when finalizing the trade.

To tamper with the object would require unlocking it, which consumes the key. Consequently, there would no longer be a key to finish the trade.

{@inject: examples/trading/contracts/escrow/sources/lock.move#struct=Locked,Key noComments}

> - The `Locked<T>` type stores the locked `obj`, the `ID` of the key that unlocks it, and its own `id`.

Reviewer: Updating to reflect the fact that the snippet now shows the dynamic field version.

```suggestion
- The `Locked<T>` type stores the `ID` of the key that unlocks it, and its own `id`. The object being locked will be added as a dynamic object field, so that it is still readable at its own ID off-chain.
```

Can we also add a link on "dynamic object field" or link to the appropriate concept?

I also left this comment on the previous PR:

> The ideal case is that we explain this progression: start with the code that stores the obj inline, and then when we need to display the object in the front-end, we explain the requirement, and the guide runs someone through updating the Locked type to make that work at that point. But let's add that as a future improvement?

Can we track that future improvement somewhere?

</details>

The `unlock` function accepts the `Locked<T>` and `Key` by value (which consumes them), and returns the underlying `T` as long as the correct key has been supplied for the lock:

<details>
<summary>
`unlock` function in `lock.move`
</summary>
{@inject: examples/trading/contracts/escrow/sources/lock.move#variable=ELockKeyMismatch noComments}
{@inject: examples/trading/contracts/escrow/sources/lock.move#fun=unlock noTitle noComments}
</details>

Together, they ensure that a lock and key cannot have existed before the lock operation, and will not exist after a successful unlock – it is single use.

:::tip Additional resources

- [Full source code](https://github.com/MystenLabs/sui/blob/705ee1ed3ce8cfadc4597c6facb6769d7dfb5896/examples/trading/contracts/escrow/sources/lock.move)
- Concepts: [Wrapped Objects](../../../concepts/versioning.mdx#wrapped-objects)

:::

### Testing Locked and Key

Move's type system guarantees that a given `Key` cannot be re-used (because `unlock` accepts it by value), but there are some properties that need to be confirmed with tests:

- A locked object can be unlocked with its key.

> - Trying to lock an object with the wrong key will fail.

Reviewer: ```suggestion
- Trying to unlock an object with the wrong key will fail.
```

{@inject: examples/trading/contracts/escrow/sources/lock.move#fun=test_coin noComments}

- All test-related functions and imports are annotated with `#[test_only]` to make sure they don't show up in the published package. This can also be done by separating tests into their own module – e.g. `lock_tests.move` – and marking that module as `#[test_only]`.
- The `test_scenario` module is used to provide access to a `&mut TxContext` in the test (necessary for creating new objects). Tests that don't need to simulate multiple transactions but still need access to a `TxContext` can use `sui::tx_context::dummy` to create a test context instead.

The first test works by creating an object to test, locking it and unlocking it – this should finish executing without aborting.
The last two lines exist to keep the Move compiler happy by cleaning up the test coin and test scenario objects, because values in Move will not be implicitly cleaned up unless they have the `drop` ability.

> {@inject: examples/trading/contracts/escrow/sources/lock.move#fun=test_lock_unlock noComments}

Reviewer: > In the actual code, we're calling `coin::burn_for_testing(coin)` -- this could probably be `coin.burn_for_testing()`.

(From previous PR).

1. The first party completes the trade by providing their locked object and the key to unlock it. Assuming all checks pass, this transfers their object to the second party and makes the second party's object available to them.

You can start by implementing steps two and three, by defining a new type to hold the escrowed object. It holds the `escrowed` object and an `id: UID` (because it's an object in its own right), but it also records the `sender` and intended `recipient` (to confirm they match when the trade happens), and it registers interest in the first party's object by recording the `ID` of the key that unlocks the `Locked<U>` that contains the object.

{@inject: examples/trading/contracts/escrow/sources/shared.move#struct=Escrow noComments singleSpace}

You also need to create a function for creating the `Escrow` object. The object is shared because it needs to be accessed by the address that created it (in case the object needs to be returned) and by the intended recipient (to complete the swap).

<details>
<summary>
`create` function in `shared.move`
</summary>

> {@inject: examples/trading/contracts/escrow/sources/shared.move#fun=create noComments}

Reviewer: Would it be possible to not include the `event::emit` calls in the snippet at this point? Because we only explain them at a later point.

<details>
<summary>
`swap` function in `shared.move`
</summary>
{@inject: examples/trading/contracts/escrow/sources/shared.move#variable=EMismatchedSenderRecipient,EMismatchedExchangeObject singleSpace noComments}
{@inject: examples/trading/contracts/escrow/sources/shared.move#fun=swap noComments noTitle}
</details>

:::tip Additional resources

- [Full source code](https://github.com/MystenLabs/sui/blob/705ee1ed3ce8cfadc4597c6facb6769d7dfb5896/examples/trading/contracts/escrow/sources/shared.move)
- Concepts: [Shared Objects](../../../concepts/object-ownership/shared)
- Concepts: [Shared Object Deletion](https://blog.sui.io/ephemeral-shared-objects/)
- Concepts: [PTBs](../../../concepts/transactions/prog-txn-blocks)

:::

### Testing

Tests for the `escrow` module will be more involved than for `lock` – you will take advantage of `test_scenario`'s ability to simulate multiple transactions from different senders, and interact with shared objects.

> You will focus on the test for a successful swap, but a link to all the tests can be found later on.

Reviewer: This scans strangely for me -- it comes from changing "We" to "You" but leaving the rest of the sentence unchanged, I think.

As with the lock test, start by creating a function to mint a test coin. You also create some constants to represent our transaction senders, `ALICE`, `BOB`, and `DIANE`.

{@inject: examples/trading/contracts/escrow/sources/shared.move#fun=test_coin noComments}

The test body starts with a call to `test_scenario::begin` and ends with a call to `test_scenario::end`. It doesn't matter which address you pass to `begin`, because you will pick one of `ALICE` or `BOB` at the start of each new transaction you write, so set it to `@0x0`:

{@inject: examples/trading/contracts/escrow/sources/shared.move#test}

> The first transaction is from `BOB` who creates a coin and locks it. You may remember the ID of the coin and the ID of the key, which you will need later, but now you transfer the locked object and the key itself to `BOB`, because this is what would happen with a real transaction: When simulating transactions in a test, you should only keep around primitive values, not whole objects, which would need to be written to chain between transactions.

Reviewer: Similar scanning issue for me here:

```suggestion
The first transaction is from `BOB` who creates a coin and locks it. You must remember the ID of the coin and the ID of the key, which you will need later, and then you transfer the locked object and the key itself to `BOB`, because this is what would happen in a real transaction: When simulating transactions in a test, you should only keep around primitive values, not whole objects, which would need to be written to chain between transactions.
```

{@inject: examples/trading/contracts/escrow/sources/shared.move#struct=EscrowCreated,EscrowSwapped,EscrowCancelled noComments}

Functions responsible for various aspects of the escrow's lifecycle emit these events. The custom indexer can then subscribe to transactions that emit these events and process only those, rather than the entire chain state:

<details>
<summary>
`emit` events included in functions from `shared.move`
</summary>
{@inject: examples/trading/contracts/escrow/sources/shared.move#use=sui::event}
{@inject: examples/trading/contracts/escrow/sources/shared.move#fun=create,swap,return_to_sender noTitle noComments}
</details>

:::tip Additional resources

- Concepts: [Events](../../../concepts/events)

:::

### Next steps

Well done. You have written the Move package! 🚀

To turn this into a complete dApp, you need to create a frontend. However, for the frontend to be updated, it has to listen to the blockchain as escrows are made and swaps are fulfilled.

To achieve this, in the next step you create an indexing service.

> ## Backend indexer {#backend}

Reviewer: (Follow-up, from previous PR):

> Explanation of the schema and indexer could also be broken up into stages -- explain the initial set-up, then how to index one kind of event end-to-end, then add subsequent events (which require less explanation), to avoid having to justify all parts of the schema up-front without it being clear what they would be used for.
>
> We could take this to the next level -- i.e. only build out the locked.move Move module, its indexing, and the UI for locking and unlocking objects before moving on to the Escrow part, but I don't think that's worthwhile -- I think it's easy to justify the interface to each high level part (Move package, Backend, API, Front-end) and then break it down at that level.
>
> This can be a follow-up.
